contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -1564,3 +1565,72 @@ smallint_radians(PG_FUNCTION_ARGS)
 
 	PG_RETURN_INT32((int32)result);
 }
+
+/* Returns the database schema name for schema-scoped objects. */
+Datum 
+object_schema_name(PG_FUNCTION_ARGS)
+{	
+	Oid object_id;
+	Oid database_id;
+	Oid	user_id = GetUserId();
+	Oid namespace_oid = InvalidOid;
+	Oid relid = InvalidOid;
+	char* namespace_name;
+	const char* schema_name;
+
+	if(PG_ARGISNULL(0))
+		PG_RETURN_NULL();
+	else
+		object_id = (Oid) PG_GETARG_INT32(0);
+
+	if(PG_ARGISNULL(1))
+		database_id = get_cur_db_id();
+	else {
+		database_id = (Oid) PG_GETARG_INT32(1);
+		user_id = GetSessionUserId();
+	}
+
+	/* if user inputs invalid database_id */
+	if (!DbidIsValid(database_id))
+		PG_RETURN_NULL();
+	
+	/* lookup namespace_oid in pg_class */
+	relid = get_rel_namespace(object_id);
+	if(OidIsValid(relid) && 
+		pg_class_aclcheck(object_id, user_id, ACL_SELECT) == ACLCHECK_OK)
+	{
+		namespace_oid = relid;
+	}
+	
+	if (!OidIsValid(namespace_oid)) /* if not found earlier */
+	{
+		/* Lookup namespace_oid in pg_constraint */
+		namespace_oid = tsql_get_constraint_nsp_oid(object_id, user_id);
+	
+	}
+	if (!OidIsValid(namespace_oid))  /* if not found earlier */
+	{
+		/* Lookup namespace_oid in pg_proc */
+		namespace_oid = tsql_get_proc_nsp_oid(object_id, user_id);
+	}
+	if (!OidIsValid(namespace_oid))  /* if not found earlier */
+	{
+		/* Lookup namespace_oid in pg_trigger */
+		namespace_oid = tsql_get_trigger_nsp_oid(object_id, user_id);
+	}
+
+	/* Find schema name from namespace_oid */
+	if (OidIsValid(namespace_oid)){
+		namespace_name = get_namespace_name(namespace_oid);
+
+		if (pg_namespace_aclcheck(namespace_oid, user_id, ACL_USAGE) != ACLCHECK_OK ||
+			database_id != get_dbid_from_physical_schema_name(namespace_name, true))
+				PG_RETURN_NULL();
+
+		schema_name = pstrdup(get_logical_schema_name(namespace_name, true));
```

Comment:
no need duplicate the string, cstring_to_text makes a copy

Reply:
Updated in latest commit.